tests/baselines/reference/mappedTypeWithAny.errors.txt

```diff
@@ -28,4 +30,30 @@ tests/cases/conformance/types/mapped/mappedTypeWithAny.ts(23,16): error TS2339:
                    ~~~~~~~~~
 !!! error TS2339: Property 'notAValue' does not exist on type 'Data'.
     }
-    
\ No newline at end of file
+    
+    // Issue #46169.
+    // We want mapped types whose constraint is `keyof T` to
+    // map over `any` differently, depending on whether `T`
+    // is constrained to array and tuple types.
+    type Arrayish<T extends unknown[]> = { [K in keyof T]: T[K] };
+    type Objectish<T extends unknown> = { [K in keyof T]: T[K] };
+    
+    // When a mapped type whose constraint is `keyof T` is instantiated,
+    // `T` may be instantiated with a `U` which is constrained to
+    // array and tuple types. When `U` is later instantiated with `any`,
+    // the result should also be some sort of array.
+    type IndirectArrayish<U extends unknown[]> = Objectish<U>;
+    
+    function bar(arrayish: Arrayish<any>, objectish: Objectish<any>, indirectArrayish: IndirectArrayish<any>) {
+        let arr: any[];
+        arr = arrayish;
+        arr = objectish;
+        ~~~
+!!! error TS2740: Type 'Objectish<any>' is missing the following properties from type 'any[]': length, pop, push, concat, and 16 more.
+        arr = indirectArrayish;
+        ~~~
+!!! error TS2322: Type 'Objectish<any>' is not assignable to type 'any[]'.

```

Comment:
My intent was that `IndirectArrayish<U>` should be a mapped object type that has a type variable constraint of `unknown[]`; it seems like that might not be the case, so I actually might need some help here.

Reply:
I think you'd need a new assignability rule for mapped types matching that pattern to accomplish that.

Reply:
I think I see what you're saying - that you need to have something recursive that checks if you're arrayish or a generic mapped type that would produce an array result.

```ts
function mappedTypeConstraintProducesArrayResult(type: Type) {
    if (isArrayType(type) || isTupleType(type)) return true;

    const typeVariable = type.flags & TypeFlags.MappedType && getHomomorphicTypeVariable(type as MappedType);
    if (typeVariable && !type.declaration.nameType) {
        const constraint = getConstraintOfTypeParameter(typeVariable);
        return !!(constraint && mappedTypeConstraintProducesArrayResult(constraint));
    }
    return false;
}
```

But `IndirectArrayish<any>` and `Objectish<any>` aren't generic mapped types, are they?

Reply:
Uhh, I was thinking of something like relating a generic `IndirectArrayish<U>` to a `U[]`. In this case, I don't think we can distinguish between a `Objectish<any>` and an `IndirectArrayish<any>` since they're exactly equivalent - the "is this constrained to arrayish types only" check you have is purely syntactic (it only cares if the mapped type was both declared homomorphic and that variable is array-constrained) - it can't pick up information from wrapping declarations or anything.

Reply:
I guess the intent here was that when we declared `IndirectArrayish<U extends unknown[]>`, we'd end up with a fresh generic mapped type whose type variable has the array constraint. When we then go to instantiate `any` for `U`, we'd check the constraint for `U`.